clients/pkg/promtail/targets/gcplog/push_translation.go

```diff
@@ -73,45 +74,46 @@ func translate(m PushMessage, other model.LabelSet, useIncomingTimestamp bool, r
 		labels[model.LabelName(lbl.Name)] = model.LabelValue(lbl.Value)
 	}
 
-	// add labels coming from scrapeconfig
+	// Add fixed labels coming from the target configuration
 	labels = labels.Merge(other)
 
-	ts := time.Now()
-
 	decodedData, err := base64.StdEncoding.DecodeString(m.Message.Data)
 	if err != nil {
 		return api.Entry{}, fmt.Errorf("failed to decode data: %w", err)
 	}
-	line := string(decodedData)
-
-	if useIncomingTimestamp {
-		var err error
-		ts, err = time.Parse(time.RFC3339, m.Message.PublishTimestamp)
-		if err != nil {
-			return api.Entry{}, fmt.Errorf("invalid publish timestamp format: %w", err)
-		}
-	}
 
 	// If the incoming request carries the tenant id, inject it as the reserved label, so it's used by the
 	// remote write client.
 	if xScopeOrgID != "" {
 		labels[lokiClient.ReservedLabelTenantID] = model.LabelValue(xScopeOrgID)
 	}
 
-	return api.Entry{
-		Labels: labels,
-		Entry: logproto.Entry{
-			Timestamp: ts,
-			Line:      line,
-		},
-	}, nil
+	// The labelset built by merging fixed target-level labels, push-specific labels and tenant ID is passed as to be merged
+	// to the resulting one of processing the actual GCP Logs entry.
+	entry, err := parseGCPLogsEntry(decodedData, labels, useIncomingTimestamp, relabelConfigs)
+	if err != nil {
+		return api.Entry{}, fmt.Errorf("failed to process logs entry: %w", err)
+	}
+
+	return entry, nil
 }
 
-// convertToLokiCompatibleLabel converts an incoming GCP Push message label to a loki compatible format. There are lables
+// convertToLokiCompatibleLabel converts an incoming GCP Push message label to a loki compatible format. There are labels
 // such as `logging.googleapis.com/timestamp`, which contain non-loki-compatible characters, which is just alphanumeric
 // and _. The approach taken is to translate every non-alphanumeric separator character to an underscore.
 func convertToLokiCompatibleLabel(label string) string {
-	// TODO: Since this is running for every incoming message, maybe it's more performant to do something
-	// like a loop over characters, and checking if it's not loki compatible, instead of a regexp.
-	return util.SnakeCase(labelToLokiCompatible.ReplaceAllString(label, "_"))
+	return util.SnakeCase(convertSeparatorCharacterToUnderscore(label))
+}
+
+func convertSeparatorCharacterToUnderscore(s string) string {
```

Comment:
Thanks for this change. :+1: 
should we lock this with some test cases? (if there isn't any?).